scalding-core/src/main/scala/com/twitter/scalding/typed/OptimizationRules.scala

```diff
@@ -0,0 +1,344 @@
+
+        def toFunction[A] = {
+          case (c: CrossPipe[a, b], f) =>
+            Binary(f(c.left), f(c.right), CrossPipe(_: TypedPipe[a], _: TypedPipe[b]))
+          case (cv@CrossValue(_, _), f) =>
+            def go[A, B](cv: CrossValue[A, B]): LitPipe[(A, B)] =
+              cv match {
+                case CrossValue(a, ComputedValue(v)) =>
+                  Binary(f(a), f(v), { (a: TypedPipe[A], b: TypedPipe[B]) =>
+                    CrossValue(a, ComputedValue(b))
+                  })
+                case CrossValue(a, v) =>
+                  Unary(f(a), CrossValue(_: TypedPipe[A], v))
+              }
+            widen(go(cv))
+          case (p: DebugPipe[a], f) =>
+            Unary(f(p.input), DebugPipe(_: TypedPipe[a]))
+          case (p: FilterKeys[a, b], f) =>
+            widen(Unary(f(p.input), FilterKeys(_: TypedPipe[(a, b)], p.fn)))
+          case (p: Filter[a], f) =>
+            Unary(f(p.input), Filter(_: TypedPipe[a], p.fn))
+          case (p: Fork[a], f) =>
+            Unary(f(p.input), Fork(_: TypedPipe[a]))
+          case (p: FlatMapValues[a, b, c], f) =>
+            widen(Unary(f(p.input), FlatMapValues(_: TypedPipe[(a, b)], p.fn)))
+          case (p: FlatMapped[a, b], f) =>
+            Unary(f(p.input), FlatMapped(_: TypedPipe[a], p.fn))
+          case (p: ForceToDisk[a], f) =>
+            Unary(f(p.input), ForceToDisk(_: TypedPipe[a]))
+          case (it@IterablePipe(_), _) =>
+            Literal.Const(it)
+          case (p: MapValues[a, b, c], f) =>
+            widen(Unary(f(p.input), MapValues(_: TypedPipe[(a, b)], p.fn)))
+          case (p: Mapped[a, b], f) =>
+            Unary(f(p.input), Mapped(_: TypedPipe[a], p.fn))
+          case (p: MergedTypedPipe[a], f) =>
+            Binary(f(p.left), f(p.right), MergedTypedPipe(_: TypedPipe[a], _: TypedPipe[a]))
+          case (src@SourcePipe(_), _) =>
+            Literal.Const(src)
+          case (p: SumByLocalKeys[a, b], f) =>
+            widen(Unary(f(p.input), SumByLocalKeys(_: TypedPipe[(a, b)], p.semigroup)))
+          case (p: TrappedPipe[a], f) =>
+            Unary(f(p.input), TrappedPipe[a](_: TypedPipe[a], p.sink, p.conv))
+          case (p: WithDescriptionTypedPipe[a], f) =>
+            Unary(f(p.input), WithDescriptionTypedPipe(_: TypedPipe[a], p.description, p.deduplicate))
+          case (p: WithOnComplete[a], f) =>
+            Unary(f(p.input), WithOnComplete(_: TypedPipe[a], p.fn))
+          case (EmptyTypedPipe, _) =>
+            Literal.Const(EmptyTypedPipe)
+          case (hg: HashCoGroup[a, b, c, d], f) =>
+            widen(handleHashCoGroup(hg, f))
+          case (CoGroupedPipe(cg), f) =>
+            widen(handleCoGrouped(cg, f))
+          case (ReduceStepPipe(rs), f) =>
+            widen(handleReduceStep(rs, f))
+        }
+      })
+
+    private def handleReduceStep[K, V1, V2](rs: ReduceStep[K, V1, V2], recurse: FunctionK[TypedPipe, LitPipe]): LitPipe[(K, V2)] =
```

Comment:
what do you think about breaking out the handle* methods in different src files? Feel it might improve readability as we won't have a ton of code in one src file.

Reply:
If we do that, they can't be private. That has binary compatibility issues and also just general API issues.

I'd prefer to keep them private.

Reply:
Hmm yeah, that's true.